db/persistent_stats_history.cc

```diff
@@ -0,0 +1,156 @@
+
+bool PersistentStatsHistoryIterator::Valid() const { return valid_; }
+
+Status PersistentStatsHistoryIterator::status() const { return status_; }
+
+void PersistentStatsHistoryIterator::Next() {
+  // increment start_time by 1 to avoid infinite loop
+  AdvanceIteratorByTime(GetStatsTime() + 1, end_time_);
+}
+
+uint64_t PersistentStatsHistoryIterator::GetStatsTime() const { return time_; }
+
+const std::map<std::string, uint64_t>&
+PersistentStatsHistoryIterator::GetStatsMap() const {
+  return stats_map_;
+}
+
+int PersistentStatsHistoryIterator::EncodeVersionKey(uint64_t now_micros,
+                                                     int size, char* buf) {
+  // make time stamp string equal in length to allow sorting by time
+  int length = snprintf(buf, size, "%010d#%s",
+                        static_cast<int>(now_micros / kMicrosInSecond),
+                        kVersionKeyString);
+  return length;
+}
+
+int PersistentStatsHistoryIterator::DecodeVersionNumber(
+    const char* timestamp_str, Iterator* iter) {
+  char version_key[100];
+  snprintf(version_key, 100, "%s#%s", timestamp_str, kVersionKeyString);
+  iter->Seek(version_key);
+  // cannot find version_number, could be a user-defined CF or internal error
+  // skip all with the same timestamp
+  if (!iter->Valid()) {
+    return -1;
+  } else {
+    // read version_number but do nothing in current version
+    int version_number =
+        static_cast<int>(std::stoull(iter->value().ToString()));
+    // reset iter to beginning
+    iter->Seek(timestamp_str);
+    return version_number;
+  }
+}
+
+int PersistentStatsHistoryIterator::EncodeKey(uint64_t now_micros,
+                                              const std::string& key, int size,
+                                              char* buf) {
+  char timestamp[kNowSecondsStringLength + 1];
+  // make time stamp string equal in length to allow sorting by time
+  snprintf(timestamp, sizeof(timestamp), "%010d",
+           static_cast<int>(now_micros / kMicrosInSecond));
+  timestamp[kNowSecondsStringLength] = '\0';
+  return snprintf(buf, size, "%s#%s", timestamp, key.c_str());
+}
+
+std::pair<uint64_t, std::string> parseKey(const Slice& key,
+                                          uint64_t start_time) {
+  std::pair<uint64_t, std::string> result;
+  std::string key_str = key.ToString();
+  std::string::size_type pos = key_str.find("#");
+  // TODO(Zhongyi): add counters to track parse failures?
+  if (pos == std::string::npos) {
+    result.first = 0;
+    result.second = "";
+  } else {
+    uint64_t parsed_time = ParseUint64(key_str.substr(0, pos));
+    // skip entries with timestamp smaller than start_time
+    if (parsed_time < start_time) {
+      result.first = 0;
+      result.second = "";
+    } else {
+      result.first = parsed_time;
+      std::string key_resize = key_str.substr(pos + 1);
+      auto it = std::find(key_resize.begin(), key_resize.end(), '\0');
+      assert(it == key_resize.end());
```

Comment:
I'm thinking that assertion may be too strong for this. It is true that RocksDB is supposed to write to this CF and now RocksDB reads from it. But writing to it might happen by another version and may even from another host. Asserting data correctness of a DB not generated by the current process sounds not right to me.

Reply:
sure, will remove the assert and see if it's possible to move the check in a unit test